ast/check.go

```diff
@@ -749,40 +750,69 @@ func (a *UnificationErrDetail) Lines() []string {
 	return lines
 }
 
-// RefErrDetail represents an undefined ref.
-type RefErrDetail struct {
-	Ref      Ref        `json:"ref"`
-	Index    int        `json:"index"`
-	Refers   types.Type `json:"refers"`
-	Referrer types.Type `json:"referrer"`
+// RefErrUnsupported describes an undefined reference error where the
+// referenced value does not support dereferencing (e.g., scalars).
+type RefErrUnsupported struct {
+	Ref  Ref        `json:"ref"`  // invalid ref
+	Pos  int        `json:"pos"`  // invalid element
+	Have types.Type `json:"have"` // referenced type
 }
 
 // Lines returns the string representation of the detail.
-func (r *RefErrDetail) Lines() []string {
-	lines := []string{}
-	lines = append(lines, fmt.Sprintf("prefix    : %v (valid)", r.Ref[:r.Index]))
-	if r.Refers != nil {
-		lines = append(lines, fmt.Sprintf("refers to : %v", r.Refers))
-	}
-	tail := r.Ref[r.Index:]
-	var str string
-	if len(tail) == 1 {
-		switch v := tail[0].Value.(type) {
-		case String:
-			str = formatString(v)
-		default:
-			str = v.String()
-		}
+func (r *RefErrUnsupported) Lines() []string {
+	lines := []string{r.Ref.String()}
+	offset := len(r.Ref[:r.Pos].String()) + 1
+	pad := strings.Repeat(" ", offset)
+	lines = append(lines, fmt.Sprintf("%v", strings.Repeat("^", len(r.Ref[:r.Pos+1].String()))))
```

Comment:
This code is ugly. I'll refactor so that it's more extensible (e.g., we could use colors instead of carets.)